File: frontends/Matlab/Postprocessing/exasim.m
function [sol,pde,mesh,master,dmd,compilerstr,runstr,res] = exasim(pde,mesh)

if iscell(pde)
    nmodels = length(pde);
    if nmodels==1
        error("Number of PDE models must be greater than 1");
    end
else
    nmodels = 1;
end

res = [];
if nmodels==1
    % search compilers and set options
    % pde = setcompilers(pde);       
    
    % generate input files and store them in datain folder
    [pde,mesh,master,dmd] = preprocessing(pde,mesh);

    % generate source codes and store them in app folder
    if pde.gencode==1
      %gencode(pde);
      kkgencode(pde);
      compilerstr = cmakecompile(pde); % use cmake to compile C++ source codes 
      %compilerstr = compilepdemodel(pde);
    end
           
    runstr = runcode(pde, 1); % run C++ code

    % get solution from output files in dataout folder
    sol = fetchsolution(pde,master,dmd, pde.buildpath + "/dataout");
    
    % get residual norms from output files in dataout folder
    if pde.saveResNorm
        fileID = fopen('dataout/out_residualnorms0.bin','r'); res = fread(fileID,'double'); fclose(fileID);
        res = reshape(res,4,[])';
    end    
else
    master = cell(nmodels,1);
    dmd = cell(nmodels,1);
    sol = cell(nmodels,1);
    res = cell(nmodels,1);
    
    nummodels = nmodels;
    mpiprocs = pde{1}.mpiprocs;
    if isfield(mesh{1}, 'interfacecondition')           
      if max(mesh{1}.interfacecondition) >= 1
        nummodels = 100 + pde{1}.mpiprocs;
        mpiprocs = mpiprocs + pde{2}.mpiprocs;        
      end
    end

    % preprocess all PDE models
    for m = 1:nmodels    
        [pde{m},mesh{m},master{m},dmd{m}] = preprocessing(pde{m},mesh{m});
    end                  
    
    if isfield(mesh{1}, 'interfacecondition')        
      [dmd{1},dmd{2},isd1,isd2]=interfacepartition(mesh{1}, dmd{1}, mesh{2}, dmd{2});
      writedmd(dmd{1}, pde{1}, isd1);
      writedmd(dmd{2}, pde{2}, isd2);
    end
          
    % generate code for all PDE models
    if pde{1}.gencode==1
      for m = 1:nmodels    
        kkgencode(pde{m});
      end      
      kkgencodeall(nmodels, pde{1}.backendpath + "/Model");
      compilerstr = cmakecompile(pde{1}, mpiprocs); % use cmake to compile C++ source codes 
      %compilerstr = compilepdemodel(pde{1}); % use cmake to compile C++ source codes 
    end
           
    runstr = runcode(pde{1}, nummodels, mpiprocs); % run C++ code
    
    % get solution from output files in dataout folder
    for m = 1:nmodels        
        sol{m} = fetchsolution(pde{m},master{m},dmd{m}, pde{m}.buildpath + "/dataout" + num2str(m));                
    end    
end

File: frontends/Julia/Postprocessing/exasim.jl
function  exasim(pde,mesh)

cdir = pwd(); ii = findlast("Exasim", cdir);
ExasimPath = cdir[1:ii[end]];
buildpath = ExasimPath * "/build";

if isa(pde, Array)
    nmodels = length(pde);      
else
    nmodels = 1;   
end
res = []
if nmodels==1
    # search compilers and set options
    pde = Gencode.setcompilers(pde);

    # generate input files and store them in datain folder
    pde, mesh, master, dmd = Preprocessing.preprocessing(pde,mesh);

    # generate source codes and store them in app folder
    Gencode.gencode(pde);

    compilerstr = Main.cmakecompile(pde); # use cmake to compile source codes 
    #compilerstr = Main.compilepdemodel(pde); # use cmake to compile source codes 

    runstr = Gencode.runcode(pde, 1);

    # get solution from output files in dataout folder
    sol = Postprocessing.fetchsolution(pde,master,dmd, buildpath * "/dataout");
    if pde.saveResNorm == 1
        fn = "dataout/out_residualnorms0.bin";
        res = reinterpret(Float64,read(fn));        
        ne = Int64(round(length(res)/4));
        res = reshape(res,(4,ne));                
        res = res';
    end    
else
    master = Array{Any, 1}(undef, nmodels);
    dmd = Array{Any, 1}(undef, nmodels);
    sol = Array{Any, 1}(undef, nmodels);
    res = Array{Any, 1}(undef, nmodels);

    # preprocess and generate code for all PDE models
    for m = 1:nmodels    
        pde[m],mesh[m],master[m],dmd[m] = generatecode(pde[m],mesh[m]);
    end        
    gencodeall(nmodels);

    if pde.usecmake==1    
        compilerstr = Main.cmakecompile(pde[1]); # use cmake to compile source codes 
        #compilerstr = "";    
        #compilerstr = Main.compilepdemodel(pde[1]); # use cmake to compile source codes 
        runstr = runcode(pde[1],nmodels);
    else
        # # compile source codes to build an executable file and store it in app folder
        compilerstr = compilecode(pde[1]);

        # run executable file to compute solution and store it in dataout folder
        runstr = runcode(pde[1],nmodels);
    end

    # get solution from output files in dataout folder
    for m = 1:nmodels        
        sol[m] = Postprocessing.fetchsolution(pde[m], master[m], dmd[m], "dataout" * string(m));
        #sol[m],~,~ = Postprocessing.getsolutions(pde[m], dmd[m]);

        if pde[m].saveResNorm
            fn = "dataout/out_residualnorms" * string(m-1) * ".bin";
            tm = reinterpret(Float64,read(fn));        
            ne = Int64(round(length(tm)/4));
            tm = reshape(tm,(4,ne));                
            res[m] = tm';
        end    
    end    
end

return sol,pde,mesh,master,dmd,compilerstr,runstr,res

end

